backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/ModularContractService.java

```diff
@@ -0,0 +1,124 @@
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.contracts.modular;
+
+import de.metas.contracts.FlatrateTermId;
+import de.metas.contracts.IFlatrateDAO;
+import de.metas.contracts.model.I_C_Flatrate_Term;
+import de.metas.contracts.model.X_C_Flatrate_Term;
+import de.metas.contracts.modular.log.ModularContractLogDAO;
+import de.metas.contracts.modular.settings.ModularContractSettings;
+import de.metas.contracts.modular.settings.ModularContractSettingsDAO;
+import de.metas.contracts.modular.settings.ModularContractType;
+import de.metas.contracts.modular.settings.ModuleConfig;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.springframework.stereotype.Service;
+
+import java.util.Objects;
+
+@Service
+public class ModularContractService
+{
+	public enum ModelAction
+	{
+		COMPLETED, REVERSED, REACTIVATED
+	}
+
+	private final ModularContractLogDAO contractLogDAO;
+	private final ModularContractHandlerFactory modularContractHandlerFactory;
+	private final ModularContractSettingsDAO modularContractSettingsDAO;
+	private final IFlatrateDAO flatrateDAO = Services.get(IFlatrateDAO.class);
+
+	public ModularContractService(
+			@NonNull final ModularContractLogDAO contractLogDAO,
+			@NonNull final ModularContractHandlerFactory modularContractHandlerFactory,
+			@NonNull final ModularContractSettingsDAO modularContractSettingsDAO)
+	{
+		this.contractLogDAO = contractLogDAO;
+		this.modularContractHandlerFactory = modularContractHandlerFactory;
+		this.modularContractSettingsDAO = modularContractSettingsDAO;
+	}
+
+	public void invokeWithModel(@NonNull final Object model, @NonNull final ModelAction action)
+	{
+		modularContractHandlerFactory.getApplicableHandlersFor(model)
+				.forEach(handler -> invokeWithModel(handler, model, action));
+	}
+
+	private <T> void invokeWithModel(@NonNull final IModularContractTypeHandler<T> handler, final @NonNull T model, final @NonNull ModelAction action)
+	{
+		handler.getContractIds(model)
+				.filter(flatrateTermId -> isApplicableContract(handler, flatrateTermId))
+				.forEach(flatrateTermId -> invokeWithModel(handler, model, action, flatrateTermId));
+	}
+
+	private <T> boolean isApplicableContract(@NonNull final IModularContractTypeHandler<T> handler, @NonNull final FlatrateTermId flatrateTermId)
+	{
+		if (!isModularContract(flatrateTermId))
+		{
+			return false;
+		}
+
+		final ModularContractSettings settings = modularContractSettingsDAO.getFor(flatrateTermId);
+		return isHandlerApplicableForSettings(handler, settings);
+	}
+
+	private boolean isModularContract(@NonNull final FlatrateTermId flatrateTermId)
+	{
+		final I_C_Flatrate_Term flatrateTerm = flatrateDAO.retrieveTerm(flatrateTermId);
+		return Objects.equals(X_C_Flatrate_Term.TYPE_CONDITIONS_ModularContract, flatrateTerm.getType_Conditions());
+	}
+
+	private <T> boolean isHandlerApplicableForSettings(@NonNull final IModularContractTypeHandler<T> handler, final ModularContractSettings settings)
```

Comment:
I would rename it to isApplicable, to avoid the repetition. It's clear from the parameter names what is applicable for what